ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/linearoptimisation/algorithms/fillers/RaUsageLimitsFiller.java

```diff
@@ -79,7 +89,26 @@ public void updateBetweenSensiIteration(LinearProblem linearProblem, FlowResult
 
     @Override
     public void updateBetweenMipIteration(LinearProblem linearProblem, RangeActionActivationResult rangeActionActivationResult) {
-        // nothing to do, we are only comparing optimal and pre-perimeter setpoints
+        rangeActions.forEach((state, rangeActionSet) -> {
+            Map<String, Integer> maxElementaryActionsPerTso = rangeActionLimitationParameters.getMaxElementaryActionsPerTso(state);
+            Map<String, Set<PstRangeAction>> pstRangeActionsPerTso = new HashMap<>();
+            rangeActionSet.stream()
+                .filter(PstRangeAction.class::isInstance)
+                .filter(rangeAction -> maxElementaryActionsPerTso.containsKey(rangeAction.getOperator()))
+                .map(PstRangeAction.class::cast)
+                .forEach(pstRangeAction -> pstRangeActionsPerTso.computeIfAbsent(pstRangeAction.getOperator(), tso -> new HashSet<>()).add(pstRangeAction));
+
+            for (String tso : maxElementaryActionsPerTso.keySet()) {
+                for (PstRangeAction pstRangeAction : pstRangeActionsPerTso.getOrDefault(tso, Set.of())) {
+                    // use pre-perimeter tap because PST's tap may be different from the initial tap after previous perimeter
+                    int initialTap = prePerimeterRangeActionSetpoints.getTap(pstRangeAction);
```

Comment:
it would probably be worth creating a test case for this

Reply:
The problem is deeper than that because the MIP model may be different for 2P Global Opt, and this is a general issue for all RA usage limits

I opened an issue (#1088) to track this and address the problem in the future

Seems quite complex though, it will be more relevant to solve all RA usage limits at once in a dedicated PR